File: app/src/main/java/com/hustunique/morii/util/DatabaseUtil.java
package com.hustunique.morii.util;

import static com.hustunique.morii.util.MyApplication.context;

import androidx.room.Room;

import com.hustunique.morii.database.AppDatabase;
import com.hustunique.morii.database.DiaryDao;
import com.hustunique.morii.database.DiaryInfo;
import com.hustunique.morii.database.DiaryWithSoundItemInfo;
import com.hustunique.morii.database.SoundItemInfo;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.Callable;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.Future;

public class DatabaseUtil {
    private static final ExecutorService exec = Executors.newCachedThreadPool();
    public static AppDatabase appDatabase = Room.databaseBuilder(context, AppDatabase.class, "app_database")
            .build();
    public static DiaryDao dao = appDatabase.diaryDao();

    public static List<DiaryWithSoundItemInfo> readDataFromRoomDataBase() {
        Future<List<DiaryWithSoundItemInfo>> future = exec.submit(() -> dao.getAllDiaryWithSoundItemInfo());
        List<DiaryWithSoundItemInfo> list = null;
        try {
            list = future.get();
        } catch (ExecutionException | InterruptedException e) {
            e.printStackTrace();
        }
        return list;
    }

    public static long insertDiaryInfo(DiaryInfo diaryInfo) {
        Future<Long> future = exec.submit(() -> dao.insertDiaryInfo(diaryInfo));
        long id = 0;
        try {
            id = future.get();
        } catch (ExecutionException | InterruptedException e) {
            e.printStackTrace();
        }
        return id;
    }

    public static void insertSoundItemInfo(SoundItemInfo info) {
        new Thread(() -> dao.insertAllSoundItemInfo(info)).start();
    }
}

File: app/src/main/java/com/hustunique/morii/util/DatabaseUtil.kt
package com.hustunique.morii.util

import com.hustunique.morii.database.SoundItemInfo
import com.hustunique.morii.database.DiaryInfo
import com.hustunique.morii.database.DiaryWithSoundItemInfo
import com.hustunique.morii.database.AppDatabase
import androidx.room.Room
import java.io.File
import java.util.concurrent.Callable
import java.util.concurrent.ExecutionException
import java.util.concurrent.Executors
import java.util.concurrent.Future

object DatabaseUtil {
    private const val TAG = "DatabaseUtil"
    private val exec = Executors.newCachedThreadPool()
    private val appDatabase = Room.databaseBuilder(
        MyApplication.context,
        AppDatabase::class.java,
        "app_database"
    )
        .build()
    private val dao = appDatabase.diaryDao()
    fun readDataFromRoomDataBase(): List<DiaryWithSoundItemInfo> {
        val future: Future<List<DiaryWithSoundItemInfo>> = exec.submit(
            Callable { dao.allDiaryWithSoundItemInfo })
        lateinit var list: List<DiaryWithSoundItemInfo>
        try {
            list = future.get()
        } catch (e: ExecutionException) {
            e.printStackTrace()
        } catch (e: InterruptedException) {
            e.printStackTrace()
        }
        return list
    }

    fun insertDiaryInfo(diaryInfo: DiaryInfo): Long {
        val future = exec.submit<Long> { dao.insertDiaryInfo(diaryInfo) }
        var id: Long = 0
        try {
            id = future.get()
        } catch (e: ExecutionException) {
            e.printStackTrace()
        } catch (e: InterruptedException) {
            e.printStackTrace()
        }
        return id
    }

    fun insertSoundItemInfo(info: SoundItemInfo) {
        Thread { dao.insertSoundItemInfo(info) }.start()
    }

    fun deleteDiary(id: Long) {
        Thread { dao.deleteInfoById(id) }.start()
    }

    fun deleteAudioFile(title: String, date: String) {
        File(getAudioFilePath(title, date)).delete()
    }

    fun getAudioFilePath(title: String, date: String): String {
        return MyApplication.externalPath + "/%s_%s.aac".format(
            title.replace('/', '_'), date
        )
    }
}